### PR TITLE
Fix S3 presign request

### PR DESCRIPTION
**Pull request by cjyclaire**
Fix the S3 presign  #996, #997 

cc: @mtdowling @xibz @jeskew

### CODE REVIEW

**Comment by xibz** [comment]
@cjyclaire - This change looks good. When modifying the .feature files, the cucumber test still pass, correct?

**Comment by cjyclaire** [comment]
@xibz  That's true.

**Comment by xibz** [comment]
@cjyclaire :shipit:

**Comment by mtdowling** [comment]
If the test was run when the change was made originally, then there's either something wrong with the test or the test is inadequate to detect the failure. How could this test have worked both before this change and after this change?

**Comment by xibz** [comment]
@mtdowling - No idea. I am confused there as well. Which is why I asked the question. However, I know the previous value `UNSIGNED-PAYLOAD` was originally used. So, the constant must be of that value. However, I think we need more testing or at least fix the testing to ensure the proper things are tested. We need a mock signature test with `UNSIGNED-PAYLOAD`.

**Comment by mtdowling** [comment]
Do these tests hit the actual service or are they mocked? We should be integration testing here to ensure it actually works.

**Comment by xibz** [comment]
These are integration tests. Either S3 is accepting the space character as valid or something else is wrong with the test itself. 

@cjyclaire - Can you confirm that you are getting expected responses from S3 when signing `UNSIGNED PAYLOAD` and sending `UNSIGNED-PAYLOAD`?

**Comment by jeskew** [comment]
Looking at the previous implementation of [`S3SignatureV4`](https://github.com/aws/aws-sdk-php/blob/a8b62d6990fd8ea97d1ae4ce370779db7fd247c5/src/Signature/S3SignatureV4.php#L53), the value of 'ContentSHA256' was never used as the presigned payload. The server must have ignore the value in a `x-amz-content-sha-256` query parameter when the canonical request was signed with a body checksum of 'UNSIGNED-PAYLOAD.'

**Comment by cjyclaire** [comment]
@jeskew make sense! Thank you to all!

**Comment by xibz** [comment]
@jeskew - A test should have thrown an error. A sufficient test should be added to prevent this from happening again. Thanks for chiming in :)!

**Comment by jeskew** [comment]
@xibz Those tests are currently failing on master.

**Comment by xibz** [comment]
@jeskew - Thanks for the help and clarification!

**Comment by cjyclaire** [comment]
It seems that cucumber test have a typo in command passed in as well, fix has been updated.

File: src/Signature/S3SignatureV4.php [hunk 1]
```diff
@@ -41,6 +41,11 @@ public function presign(
                 'X-Amz-Content-Sha256',
                 $this->getPresignedPayload($request)
             );
+        } else {
+            $request = $request->withHeader(
```
Comment (jeskew) [inline src/Signature/S3SignatureV4.php]: This is a no-op. You're assigning the current value of the header to the be the new value of the header.

Comment (cjyclaire) [inline src/Signature/S3SignatureV4.php]: I see, are you suggesting it is the "ContentSHA256" causing the trouble only?

Comment (jeskew) [inline src/Signature/S3SignatureV4.php]: I believe so. The test must be discarding the supplied value and using what's returned by `getPresignedPayload` for both scenarios.

**Comment by cjyclaire** [comment]
Fix updated

**Comment by xibz** [comment]
These changes look good, :shipit:

**Pull request merged**